xray/run.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import shlex
+import subprocess
+from analyze import analyze
+from enum import Enum
+from pathlib import Path
+
+WG_IFC_NAME = "xraywg1"
+
+
+def run_command(cmd, capture_output=False):
+    args = shlex.split(cmd)
+    run = subprocess.run(args, capture_output=capture_output)
+    return (run.stdout, run.stderr)
+
+
+def get_csv_name(wg, test_type, count):
+    return f"results/xray_metrics_{wg.lower()}_{test_type}_{count}.csv"
+
+
+def get_pcap_name(wg, test_type, count):
+    return f"results/{WG_IFC_NAME}_{wg.lower()}_{test_type}_{count}.pcap"
+
+
+class Wireguard(Enum):
+    NepTUN = 1
+    WgGo = 2
+    Native = 3
+    BoringTun = 4
+
+    def from_str(s):
+        if s is None or s.lower() == "neptun":
+            return Wireguard.NepTUN
+        elif s is not None and s.lower() == "wggo":
+            return Wireguard.WgGo
+        elif s is not None and s.lower() == "native":
+            return Wireguard.Native
+        elif s is not None and s.lower() == "boringtun":
+            return Wireguard.BoringTun
+        else:
+            raise Exception(f"{s} is not a valid wireguard type")
+
+
+def setup_wireguard(wg, build_neptun):
+    if wg == Wireguard.Native:
+        run_command(f"sudo ip link add dev {WG_IFC_NAME} type wireguard")
+    elif wg == Wireguard.WgGo:
+        wggo = (
+            run_command("which wireguard", capture_output=True)[0]
+            .strip()
+            .decode("utf-8")
+        )
+        run_command(f"sudo {wggo} {WG_IFC_NAME}")
+    elif wg == Wireguard.BoringTun:
+        run_command(f"sudo ../target/release/boringtun-cli {WG_IFC_NAME}")
+    else:
+        if build_neptun:
+            run_command(f"cargo run --release -p neptun-cli -- {WG_IFC_NAME}")
+        else:
+            run_command(f"sudo ../target/release/neptun-cli {WG_IFC_NAME}")
+    run_command(f"sudo ip link set dev {WG_IFC_NAME} up")
+    run_command(
+        f"sudo ip link set dev {WG_IFC_NAME} multicast off"
+    )  # Not strictly necessary but keeps the pcaps a bit cleaner
+
+
+def start_tcpdump(pcap_name):
+    return subprocess.Popen(
+        ["sudo", "tcpdump", "-ni", "any", "-w", pcap_name],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+
+def run_xray(wg, test_type, count):
+    run_command(
+        f"cargo run --release -- {wg} {test_type} {count} {get_csv_name(wg, test_type, count)}"
+    )
+
+
+def stop_tcpdump(tcpdump):
+    run_command(f"kill {tcpdump.pid}")
+
+
+def destroy_wireguard(wg):
+    if wg == Wireguard.NepTUN:
+        run_command("killall neptun-cli")
```

Comment:
While testing, I ended up with the `neptun-cli` processes hanging, and manually running this (SIGTERM) didn't kill those hanging processes, only SIGKILL worked. So I'd suggest running `killall -9 neptun-cli` here instead.

Reply:
same below for the Wireguard.BoringTun` case